source/agora/consensus/validation/Block.d

```diff
@@ -219,7 +224,59 @@ public string isInvalidReason (const ref Block block, Height prev_height,
             file, line, block.header.validators.length);
         return "Block: Invalid schnorr signature";
     }
-    return null;
+    return validateBlockTimestamp(prev_timestamp, block.header.timestamp, curr_timestamp, block_timestamp_tolerance_secs);
+}
+
+/*******************************************************************************
+
+    Check the validity of the timestamp for a block
+
+    new_block_ts is valid if and only if
+    prev_block_ts < new_block_ts <= curr_timestamp + block_timestamp_tolerance_secs
+
+    Params:
+        prev_block_ts = the timestam of the block preceding the new block
+        new_block_ts  = the timestamp of the block we are trying to validate
+        curr_timestamp = the current clock timestamp
+        block_timestamp_tolerance_secs = the proposed block timestamp should be less
+                than curr_timestamp + block_timestamp_tolerance_secs
+
+    Returns:
+        `null` if the new_block_ts is valid, otherwise a string explaining
+        the reason it is invalid.
+
+*******************************************************************************/
+
+public string validateBlockTimestamp(ulong prev_block_ts, ulong new_block_ts, ulong curr_timestamp,
+        uint block_timestamp_tolerance_secs) @safe nothrow
+{
+    string res;
+    try
+    {
+        import std.conv : to;
+        import std.datetime.systime : SysTime, unixTimeToStdTime;
+        log.trace("validateBlockTimestamp():\nnew block timestamp: {}\ncurrent timestamp: {}\ntolerance in seconds: {}",
+                    to!string(SysTime(unixTimeToStdTime(new_block_ts))),
+                    to!string(SysTime(unixTimeToStdTime(curr_timestamp))),
+                    block_timestamp_tolerance_secs);
```

Comment:
I would remove this `log.trace`, as the validation routines are made to be lightweight and  this is going to cause some noise

Reply:
I will remove it, but it seems isInvalidReason is all over littered with trace logs right now.